src/poliastro/twobody/orbit.py

```diff
@@ -1416,16 +1417,24 @@ def sample(self, values=100, *, min_anomaly=None, max_anomaly=None):
         else:
             nu_values = self._sample_open(values, min_anomaly, max_anomaly)
 
-        # FIXME: Refactor this insanity to remove a lot of back and forth
-        # 1. Convert to classical to extract nu values (either 1 or 3 are no ops)
-        # 2. Sample values of nu -> M -> times
-        # 3. Convert to cartesian to call the farnocchia propagator harcoded here (either 1 or 3 are no ops)
-        # 4. For each time, convert to classical to retain "slow" elements
-        # 5. For each time, compute nu0 -> M0 (same value always)
-        # 6. For each time, t -> M -> nu
-        # 7. For each time, convert classical to cartesian and assemble orbit
-        time_values = time.TimeDelta(self._generate_time_values(nu_values))
-        cartesian = propagate(self, time_values)
+        n = nu_values.shape[0]
+        rr, vv = coe2rv_many(
+            np.full(n, self.attractor.k.to(u.m ** 3 / u.s ** 2).value),
+            np.full(n, self.p.to(u.m).value),
+            np.full(n, self.ecc.value),
+            np.full(n, self.inc.to(u.rad).value),
+            np.full(n, self.raan.to(u.rad).value),
+            np.full(n, self.argp.to(u.rad).value),
+            nu_values.to(u.rad).value,
+        )
+
+        # Add units
+        rr = rr * u.m
+        vv = vv * u.m / u.s
```

Comment:
The doctest in line 1405 fails because `rr` now is in `m` instead of `km`. I'd rather convert this to `km` than to change the docstring, because it's what most propagators return already:

https://github.com/poliastro/poliastro/blob/bdd0dd5bc7936774da8ca9d494f914888a761877/src/poliastro/twobody/propagation.py#L159-L161

```suggestion
        rr = (rr << u.m).to(u.km)
        vv = (vv << (u.m / u.s)).to(u.km / u.s)
```